universe/Universe.cpp

```diff
@@ -1660,6 +1660,17 @@ void Universe::ForgetKnownObject(int empire_id, int object_id) {
     objects.Remove(object_id);
 }
 
+void Universe::GrantShipDesignKnowledge(int empire_id, int object_id) {
+    std::shared_ptr<const Ship> ship = GetShip(object_id);
+    if (!ship) return;
+    int design_id = ship->DesignID();
+    if (design_id == ShipDesign::INVALID_DESIGN_ID) {
+        ErrorLogger() << "GrantShipDesignKnowledge got invalid design id for ship with id " << object_id;
+    } else {
+        m_empire_known_ship_design_ids[empire_id].insert(design_id);
```

Comment:
> Pondering if it's worth calling Universe::SetEmpireKnowledgeOfShipDesign instead of re-implementing some of the checks here.. You'd also get some extra detailed error messages instead of silently returning if the ship doesn't exist, 

Hmm, well, It looks to me like Universe::SetEmpireKnowledgeOfShipDesign  *does NOT* check existence of the ship (it never even gets the ship_id), and anyways it's not an error in the current case if there is no ship for the object_id; quite often the object would be a fleet or planet or whatnot.  That's why the comment for this method notes "If \a object_id is a ship, ..."

> and would get a check that empire_id isn't ALL_EMPIRES, which probably should be checked for.

if empire_id were ALL_EMPIRES, then in both places that this current method is called from it would have already caused some complications with the object visibility maps; if there were a chance in these constructions for the empire_id to be bad then waiting for Universe::SetEmpireKnowledgeOfShipDesign would be rather late to check it.  

Although relying on Universe::SetEmpireKnowledgeOfShipDesign  could let us save a few lines of code, it would be at a cost -- by doing the INVALID_SHIP_DESIGN check here we are able to give a more informative error message, so I think it's a little better the way it is now.

> Also, the name is oddly inconsistent with that function's naming... Maybe instead it could be SetEmpireKnowledgeOfDesignOfShip ?  And perhaps change the parameter name to ship_id instead of object_id ?

As for the parameter name, since we expect that it often will NOT be a ship, it seems to me better to leave it as object_id.  As for the method name, if the current one strikes you as discordant with SetEmpireKnowledgeOfShipDesign, I myself would prefer to use HandleShipDesignKnowledge.   But if you definitely would prefer SetEmpireKnowledgeOfDesignOfShip then just say so and we'll go with that.